helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java

```diff
@@ -0,0 +1,132 @@
+package org.apache.helix.controller.changedetector;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import org.apache.helix.HelixConstants;
+import org.apache.helix.HelixException;
+import org.apache.helix.HelixProperty;
+import org.apache.helix.controller.dataproviders.ResourceControllerDataProvider;
+
+public class ResourceChangeDetector implements ChangeDetector {
+
+  private ResourceChangeCache _oldCache; // cache snapshot for previous pipeline run
+  private ResourceChangeCache _newCache; // cache snapshot for this pipeline run
+  private Map<String, ? extends HelixProperty> _oldPropertyMap;
+  private Map<String, ? extends HelixProperty> _newPropertyMap;
+
+  /**
+   * Compare the underlying HelixProperty objects and produce a collection of names of changed
+   * properties.
+   * @return
+   */
+  private Collection<String> getChangedItems() {
+    Collection<String> changedItems = new HashSet<>();
+    _oldPropertyMap.forEach((name, property) -> {
+      if (_newPropertyMap.containsKey(name) && !property.equals(_newPropertyMap.get(name))) {
```

Comment:
That was an option I considered, but here are the reasons I chose to use HelixProperty:

- It is possible that Stats' version has changed without any modification to the actual content of the ZNode. In this case, it's sort of like a false alarm in that we do not have to react to it. So we need to look at the actual content.
- HelixProperty.equals() actually is just a wrapper for ZNRecord.equals(). I wanted this component to be agnostic to the actual low-level data format of the metadata store. See:
`  @Override
  public boolean equals(Object obj) {
    if (obj == null) {
      return false;
    }
    if (obj instanceof HelixProperty) {
      HelixProperty that = (HelixProperty) obj;
      if (that.getRecord() != null) {
        return that.getRecord().equals(this.getRecord());
      }
    }
    return false;
  }`

Reply:
The only concern here will be memory usage and GC. Every cache is doing deep copy of existing  pipeline cache content. That could increase existing memory usage.

I am thinking about that what will be the bad effect of false alarm? Will that trigger complete recomputation or delta computation. Delta computation could impact the pipeline speed but this computation suppose to be fast, right?

Reply:
In the new constraint-based rebalancer, it may trigger another round of asynchronous, long calculation ("Global Baseline Calculation"), which is supposed to be a slow, compute-intensive process. The reason we want to have Cluster Change Detector in the first place was to minimize the need for such expensive computation.

@jiajunwang is my understanding correct?

As for memory usage, yes, @jiajunwang and I have held informal discussions, but we think that the current controller pipeline occupies the memory space for 2 rounds of caches anyways (because the lifecycle of the cache spans 2 pipeline runs), so we concluded that it shouldn't be too much worse.

Reply:
The version based check has been done in the cache refresh. If we can somehow leverage that, we will be able to avoid any duplicate check and optimize the performance.
But doing that now will be too much change. I would prefer to finish the current version with property comparing and leave a TODO here. Anyway, we will run a performance evaluation and decide if the optimization is required now.

Reply:
ACK. I am totally fine with that when you understand there could be a memory optimization. For short term, we can implement it as a simple way.

Reply:
Another point we need to discuss is - is it possible to have different version numbers yet have the content unchanged? In that case, we don't want a rebalance triggered, correct?

Reply:
Discussed @jiajunwang . So yes - we need to compare values so `HelixProperty.equals()` is needed :)